include/mata/utils/partition-relation-pair.hh

```diff
@@ -0,0 +1,1449 @@
+    // in the given initial partition to detect duplicates
+    // and to detect unused states
+    std::vector<char> used;
+    used.insert(used.end(), numOfStates, false);
+    
+    // initialization of the m_states vector
+    m_states.insert(m_states.end(), numOfStates, 0);
+    
+    // creating partition using given initial vector of vectors
+    size_t numOfBlocks = partition.size();
+    // iterating through initial partition blocks
+    for(size_t blockIdx = 0; blockIdx < numOfBlocks; ++blockIdx)
+    {
+        assert(!partition[blockIdx].empty() &&
+        "Partition class cannot be empty.");
+        
+        // iterating through one partition block
+        for(auto state : partition[blockIdx])
+        {
+            assert(state < numOfStates &&
+            "Invalid state name detected while creating" 
+            "a partition relation pair.");
+            assert(!used[state] && 
+            "Partition could not be created."
+            "Duplicate occurence of a state");
+            
+            used[state] = true;
+            
+            // creating a corresponding BlockItem
+            m_states[state] = m_blockItems.size();
+            m_blockItems.push_back({.state = state, .blockIdx = blockIdx});
+            
+        }
+        
+        // first and last states of the block will be used to create
+        // a corresponding node
+        State first = partition[blockIdx].front();
+        State last = partition[blockIdx].back();
+
+        // creating a corresponding block and node
+        m_nodes.push_back({.first = m_states[first], .last = m_states[last]});
+        m_blocks.push_back({.nodeIdx = blockIdx});
+    }
+    
+    // we need to detect whether there is a state which has not be used
+    // to create an additional partition block
+    bool allStatesUsed = true;
+    
+    // first and last unused states will surround a contiguous subvector
+    // of BlockItems
+    State first = 0;
+    State last = 0;
+    
+    // iterating through the vector of flags saying which states has been seen
+    for(State state = 0; state < numOfStates; ++state)
+    {
+        // if a state has been already seen and processed, 
+        // there is no need to add it to the additional block
+        if(used[state])
+        {
+            continue;
+        }
+        
+        // if there is at least one unused state, we need to 
+        // create an additional block
+        if(allStatesUsed)
+        {
+            allStatesUsed = false;
+            first = state;
+            ++numOfBlocks;
+        }
+        
+        // creating the new BlockItem
+        last = state;
+        m_states[state] = m_blockItems.size();
+        m_blockItems.push_back({.state = state, .blockIdx = numOfBlocks-1});
+    }
+    
+    // creating a new block and node if there was an unused state
+    if(!allStatesUsed)
+    { 
+        m_nodes.push_back({.first = m_states[first], .last = m_states[last]});
```

Comment:
Thanks for your suggestion. Since I have replaced `push_back` by `emplace_back` when constructing an object/structure in place, I do not use designated initializers anymore.

I have replaced `m_nodes.push_back({.first = m_states[first], .last = m_states[last]});` by `m_nodes.emplace_back(m_states[first], m_states[last]);` and I have implemented the constructor `Node(size_t fst, size_t lst)` to make it work properly. I have also made similar fixes in context of other occurrences.

Reply:
Sounds good.